### PR TITLE
eventually: new 0.5.0 API

### PR DESCRIPTION
**Pull request by ar3s3ru**

### CODE REVIEW

File: eventually-new/src/aggregate.rs [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+    /// to apply the [`State`] mutation based on the current representation of the State.
+    ///
+    /// [`Event`]: trait.Aggregate.html#associatedtype.Event
+    /// [`State`]: trait.Aggregate.html#associatedtype.State
+    /// [`Command`]: trait.Aggregate.html#associatedtype.Command
+    async fn handle(
+        &mut self,
+        state: &Self::State,
+        command: Self::Command,
+    ) -> Result<Option<Events<Self::DomainEvent>>, Self::HandleError>;
```
Comment (kurnevsky) [inline eventually-new/src/aggregate.rs:69]: What is the difference between None and empty vector?

Comment (ar3s3ru) [reply eventually-new/src/aggregate.rs:69]: No difference, only saving a `Vec` allocation :smile:

Comment (ar3s3ru) [reply eventually-new/src/aggregate.rs:69]: I'm aware of this "ambiguity", still wondering whether this is a good API design choice or not... What do you think?

Comment (kurnevsky) [reply eventually-new/src/aggregate.rs:69]: > No difference, only saving a Vec allocation

But as far as I remember `Vec::new()` doesn't allocate memory.

> What do you think?

I think I'd prefer unambiguous API. But I'm just passing by... :)

Comment (ar3s3ru) [reply eventually-new/src/aggregate.rs]: I appreciate the feedback, the ambiguity is now resolved :smile:

**Comment by adtennant** [comment]
Do you have a rough timeline for releasing this?

Is there anything that you would welcome collaboration on?

**Comment by ar3s3ru** [comment]
Hey @adtennant, don't worry about the tone of the message (which was perfectly fine) 😄 

My bad for not updating the PR or linking it to a tracking issue, lately I've been more focused on [the Go version](https://github.com/get-eventually/go-eventually) than the Rust one.

Initially, my idea for v0.5.0 was to have:
1. A complete redesign of the `AggregateRoot` trait, using it more similarly to a simple Entity with methods instead of using the `Handle` method with a `Command` enum. Essentially, hiding the Event Sourcing part away from the user.
2. Redesign the `EventStore` trait, splitting it in multiple sub-traits (Interface Segregation anyone? 😅) and use `async-trait` for the time being.
3. Remove the multi-crate approach for the main `eventually` crate, and just have a monorepo with the core library and the extension implementations (event stores, telemetry layer, etc.)

While 2. and 3. are mostly done (although not in this PR, I've admittedly rewritten `eventually` from scratch at least 3 times, lol), I've actually changed my mind on 1. (which was the most "controversial" -- for me, at least -- decision on the new version): I like the "functional" aspect of this implementation, so I'd like to keep it as such.

Next steps for the new version:
1. Definitely create a project/milestone to track the progress
2. Scope the requirements for the new API
3. Create the issues
4. Get to work

If you're interested in help, I would suggest to move the conversation in our Discord server (which looks a bit abandoned now) to keep the feedback cycle short 😄

File: eventually-new/src/aggregate.rs (hunk 1 above)
Comment (danieleades) [reply eventually-new/src/aggregate.rs]: ah this was exactly my question in #157!

File: Cargo.toml [hunk 2]
```diff
@@ -1,9 +1,4 @@
 [workspace]
 members = [
-    "eventually",
-    "eventually-core",
-    "eventually-postgres",
-    "eventually-redis",
-    "eventually-util",
-    "eventually-app-example"
+    "eventually"
```
Comment (danieleades) [inline Cargo.toml:3]: do you need a virtual manifest if you only have one member?

File: eventually/Cargo.toml [hunk 3]
```diff
@@ -12,19 +12,16 @@ categories = ["rust-patterns", "web-programming", "asynchronous", "data-structur
...
-# hasn't been specified in features?
-tracing = ["eventually-core/with-tracing", "eventually-util/with-tracing", "eventually-postgres/with-tracing"]
-postgres = ["eventually-postgres"]
-full = ["serde", "tracing", "postgres"]
 
 [dependencies]
-eventually-core = { version = "0.4.0", path = "../eventually-core" }
-eventually-util = { version = "0.4.0", path = "../eventually-util" }
-eventually-postgres = { version = "0.3.0", path = "../eventually-postgres", optional = true }
+anyhow = "1.0.38"
```
Comment (danieleades) [inline eventually/Cargo.toml:17]: it's a little unusual to use `anyhow` in a library- normally you'd prefer `thiserror` and reserve `anyhow` for binary crates. Can you explain the design choice? (i'm perfectly prepared to be convinced there's a good reason!)

**Comment by danieleades** [comment]
would love to contribute to this! how can i help?

**Comment by ar3s3ru** [comment]
Hey @danieleades, as I already said in https://github.com/get-eventually/eventually-rs/issues/133#issuecomment-912804263 I think I'll need to start planning some work for this new version, starting from the issues that require some modeling (e.g. Aggregate Roots).

I would really need implementing some of these things, would you be up for helping? 😄

**Comment by danieleades** [comment]
> Hey @danieleades, as I already said in https://github.com/get-eventually/eventually-rs/issues/133#issuecomment-912804263 I think I'll need to start planning some work for this new version, starting from the issues that require some modeling (e.g. Aggregate Roots).
> 
> I would really need implementing some of these things, would you be up for helping? 😄 

Of course! I'm having a ball playing around with this code base. If you could tell me what problems you're trying to solve, I'd love to help (if I can!).

**Comment by ar3s3ru** [comment]
Closing this PR in favor of smaller, targeted PRs.

You can follow the progress here: https://github.com/get-eventually/eventually-rs/projects/1  
I'll convert some of the cards to issues for contributors help soon 👍🏻 

I'm also looking for reviewers, if that interests you!

Thanks for now ❤️

**Pull request closed without merging**